File: Unity/Assets/VectorWar/GGPO.cpp
#include <IUnityInterface.h>

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool BeginGameDelegate(string name);
typedef bool (*BeginGameDelegate)(const char* name);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool AdvanceFrameDelegate(int flags);
typedef bool (*AdvanceFrameDelegate)(int flags);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool LoadGameStateDelegate(byte[] buffer);
typedef bool (*LoadGameStateDelegate)(char* buffer, int length);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate void FreeBufferDelegate(byte[] buffer);
typedef void (*FreeBufferDelegate)(char* buffer, int length);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool OnEventDelegate(int code, int player, int synchronizingcount, int synchronizingtotal, int disconnect_timeout, int frames_ahead);
typedef bool (*OnEventDelegate)(int code, int player, int synchronizingcount, int synchronizingtotal, int disconnect_timeout, int frames_ahead);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool LogGameStateDelegate(string filename, byte[] buffer);
typedef bool (*LogGameStateDelegate)(const char* filename, byte* buffer, int size);

//[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
//public delegate bool SaveGameStateDelegate(ref byte[] buffer, ref int len, ref int checksum, int frame);
typedef bool (*SaveGameStateDelegate)(&byte* buffer, int& len, int& cchecksum, int& frame);

vector<GGPOSession> sessions;
/*
 * The GGPOSessionCallbacks structure contains the callback functions that
 * your application must implement.  GGPO.net will periodically call these
 * functions during the game.  All callback functions must be implemented.
 */
typedef struct {
	/*
	 * begin_game callback - This callback has been deprecated.  You must
	 * implement it, but should ignore the 'game' parameter.
	 */
	bool(__cdecl* begin_game)(const char* game);

	/*
	 * save_game_state - The client should allocate a buffer, copy the
	 * entire contents of the current game state into it, and copy the
	 * length into the *len parameter.  Optionally, the client can compute
	 * a checksum of the data and store it in the *checksum argument.
	 */
	bool(__cdecl* save_game_state)(unsigned char** buffer, int* len, int* checksum, int frame);

	/*
	 * load_game_state - GGPO.net will call this function at the beginning
	 * of a rollback.  The buffer and len parameters contain a previously
	 * saved state returned from the save_game_state function.  The client
	 * should make the current game state match the state contained in the
	 * buffer.
	 */
	bool(__cdecl* load_game_state)(unsigned char* buffer, int len);

	/*
	 * log_game_state - Used in diagnostic testing.  The client should use
	 * the ggpo_log function to write the contents of the specified save
	 * state in a human readible form.
	 */
	bool(__cdecl* log_game_state)(char* filename, unsigned char* buffer, int len);

	/*
	 * free_buffer - Frees a game state allocated in save_game_state.  You
	 * should deallocate the memory contained in the buffer.
	 */
	void(__cdecl* free_buffer)(void* buffer);

	/*
	 * advance_frame - Called during a rollback.  You should advance your game
	 * state by exactly one frame.  Before each frame, call ggpo_synchronize_input
	 * to retrieve the inputs you should use for that frame.  After each frame,
	 * you should call ggpo_advance_frame to notify GGPO.net that you're
	 * finished.
	 *
	 * The flags parameter is reserved.  It can safely be ignored at this time.
	 */
	bool(__cdecl* advance_frame)(int flags);

	/*
	 * on_event - Notification that something has happened.  See the GGPOEventCode
	 * structure above for more information.
	 */
	bool(__cdecl* on_event)(GGPOEvent* info);
} GGPOSessionCallbacks;

public static int ggpo_start_session(BeginGameDelegate begin_game,
	AdvanceFrameDelegate advance_frame,
	LoadGameStateDelegate load_game_state,
	FreeBufferDelegate free_buffer,
	OnEventDelegate on_event,
	LogGameStateDelegate log_game_state,
	SaveGameStateDelegate save_game_state,
	string game, int num_players, int input_size, int localport) {

	auto cb = new GGPOSessionCallbacks();
	cb->on_event = on_event

	auto *session = (GGPOSession*)new Peer2PeerBackend(cb,
		game,
		localport,
		num_players,
		input_size);


	return 0;
}

File: Unity/Assets/VectorWar/GGPO.cs
//using System;

//[Serializable]
//public struct GGPOPlayer {
//    public int size;
//    public GGPOPlayerType type;
//    public int player_num;
//    public string ip_address;
//    public short port;
//}

//public class GGPOSessionCallbacks {
//    public GGPO.BeginGameDelegate begin_game;

//    public GGPO.AdvanceFrameDelegate advance_frame;

//    public GGPO.LoadGameStateDelegate load_game_state;

//    public GGPO.FreeBufferDelegate free_buffer;

//    public GGPO.OnEventDelegate on_event;

//    public GGPO.LogGameStateDelegate log_game_state;

//    public GGPO.SaveGameStateDelegate save_game_state;
//}

//public class GGPONetworkStats {
//    public int send_queue_len;
//    public int recv_queue_len;
//    public int ping;
//    public int kbps_sent;
//    public int local_frames_behind;
//    public int remote_frames_behind;
//}

//public static class GGPO {
//    public const int GGPO_OK = 0;
//    public const int GGPO_INVALID_HANDLE = -1;

//    public const int GGPO_ERRORCODE_SUCCESS = 0;
//    public const int GGPO_ERRORCODE_GENERAL_FAILURE = -1;
//    public const int GGPO_ERRORCODE_INVALID_SESSION = 1;
//    public const int GGPO_ERRORCODE_INVALID_PLAYER_HANDLE = 2;
//    public const int GGPO_ERRORCODE_PLAYER_OUT_OF_RANGE = 3;
//    public const int GGPO_ERRORCODE_PREDICTION_THRESHOLD = 4;
//    public const int GGPO_ERRORCODE_UNSUPPORTED = 5;
//    public const int GGPO_ERRORCODE_NOT_SYNCHRONIZED = 6;
//    public const int GGPO_ERRORCODE_IN_ROLLBACK = 7;
//    public const int GGPO_ERRORCODE_INPUT_DROPPED = 8;
//    public const int GGPO_ERRORCODE_PLAYER_DISCONNECTED = 9;
//    public const int GGPO_ERRORCODE_TOO_MANY_SPECTATORS = 10;
//    public const int GGPO_ERRORCODE_INVALID_REQUEST = 11;

//    public const int GGPO_EVENTCODE_CONNECTED_TO_PEER = 1000;
//    public const int GGPO_EVENTCODE_SYNCHRONIZING_WITH_PEER = 1001;
//    public const int GGPO_EVENTCODE_SYNCHRONIZED_WITH_PEER = 1002;
//    public const int GGPO_EVENTCODE_RUNNING = 1003;
//    public const int GGPO_EVENTCODE_DISCONNECTED_FROM_PEER = 1004;
//    public const int GGPO_EVENTCODE_TIMESYNC = 1005;
//    public const int GGPO_EVENTCODE_CONNECTION_INTERRUPTED = 1006;
//    public const int GGPO_EVENTCODE_CONNECTION_RESUMED = 1007;

//    public delegate bool BeginGameDelegate(string name);

//    public delegate bool AdvanceFrameDelegate(int flags);

//    public delegate bool LoadGameStateDelegate(byte[] buffer);

//    public delegate void FreeBufferDelegate(byte[] buffer);

//    public delegate bool OnEventDelegate(int code, int player, int synchronizingcount, int synchronizingtotal, int disconnect_timeout, int frames_ahead);

//    public delegate bool LogGameStateDelegate(string filename, byte[] buffer);

//    public delegate bool SaveGameStateDelegate(ref byte[] buffer, ref int len, ref int checksum, int frame);

//    // DLL
//    public static int ggpo_start_session(BeginGameDelegate begin_game,
//            AdvanceFrameDelegate advance_frame,
//            LoadGameStateDelegate load_game_state,
//            FreeBufferDelegate free_buffer,
//            OnEventDelegate on_event,
//            LogGameStateDelegate log_game_state,
//            SaveGameStateDelegate save_game_state,
//            string game, int num_players, int input_size, int localport) {
//        return 0;
//    }

//    public static int ggpo_start_session(GGPOSessionCallbacks cb, string game, int num_players, int input_size, int localport) {
//        return ggpo_start_session(cb.begin_game,
//            cb.advance_frame,
//            cb.load_game_state,
//            cb.free_buffer,
//            cb.on_event,
//            cb.log_game_state,
//            cb.save_game_state, game, num_players, input_size, localport);
//    }

//    public static int ggpo_start_spectating(BeginGameDelegate begin_game,
//            AdvanceFrameDelegate advance_frame,
//            LoadGameStateDelegate load_game_state,
//            FreeBufferDelegate free_buffer,
//            OnEventDelegate on_event,
//            LogGameStateDelegate log_game_state,
//            SaveGameStateDelegate save_game_state,
//            string game, int num_players, int input_size, int localport, string host_ip, int host_port) {
//        return 0;
//    }

//    public static int ggpo_start_spectating(GGPOSessionCallbacks cb,
//            string game,
//            int num_players,
//            int input_size,
//            int local_port,
//            string host_ip,
//            int host_port) {
//        return ggpo_start_spectating(cb.begin_game,
//            cb.advance_frame,
//            cb.load_game_state,
//            cb.free_buffer,
//            cb.on_event,
//            cb.log_game_state,
//            cb.save_game_state, game, num_players, input_size, local_port, host_ip, host_port);
//    }

//    public static void ggpo_set_disconnect_notify_start(int ggpo, int v) {
//        throw new NotImplementedException();
//    }

//    public static void ggpo_set_disconnect_timeout(int ggpo, int v) {
//        throw new NotImplementedException();
//    }

//    public static int ggpo_synchronize_input(int ggpo, out byte[] inputs, out int disconnect_flags) {
//        throw new NotImplementedException();
//    }

//    internal static int ggpo_add_local_input(int ggpo, int local_player_handle, byte[] inputs) {
//        throw new NotImplementedException();
//    }

//    internal static void ggpo_close_session(int ggpo) {
//        throw new NotImplementedException();
//    }

//    internal static void ggpo_idle(int ggpo, int time) {
//        throw new NotImplementedException();
//    }

//    internal static int ggpo_add_player(int ggpo, GGPOPlayer player, out int handle) {
//        throw new NotImplementedException();
//    }

//    internal static int ggpo_disconnect_player(int ggpo, int handle) {
//        throw new NotImplementedException();
//    }

//    internal static void ggpo_set_frame_delay(int ggpo, int handle, int frame_delay) {
//        throw new NotImplementedException();
//    }

//    internal static void ggpo_advance_frame(int ggpo) {
//        throw new NotImplementedException();
//    }

//    internal static void ggpo_log(int ggpo, string v) {
//        throw new NotImplementedException();
//    }

//    public static bool GGPO_SUCCEEDED(int result) {
//        return result == GGPO_OK;
//    }

//    public static void ggpo_get_network_stats(int ggpo, int p, out GGPONetworkStats stats) {
//        stats = new GGPONetworkStats();
//        ggpo_get_network_stats(ggpo, p,
//            out stats.send_queue_len,
//            out stats.recv_queue_len,
//            out stats.ping,
//            out stats.kbps_sent,
//            out stats.local_frames_behind,
//            out stats.remote_frames_behind
//        );
//    }

//    public static void ggpo_get_network_stats(int ggpo, int p,
//        out int send_queue_len,
//        out int recv_queue_len,
//        out int ping,
//        out int kbps_sent,
//        out int local_frames_behind,
//        out int remote_frames_behind
//    ) {
//        throw new NotImplementedException();
//    }
//}
